src/main/java/org/myrobotlab/service/abstracts/AbstractMotorController.java

```diff
@@ -1,17 +1,224 @@
+  Double minY = -1.0;
+  Double maxY = 1.0;
+  
+  // min max of input
+  Double minInput = null;
+  Double maxInput = null;
+
+  // min max of output
+  Double minOutput = null;
+  Double maxOutput = null;
 
   public AbstractMotorController(String reservedKey) {
     super(reservedKey);    
   }
+  
+  // setting default map values
+  public void map(double minX, double maxX, double minY, double maxY) {
+    this.minX = minX;
+    this.maxX = maxX;
+    this.minY = minY;
+    this.maxY = maxY;
+    broadcastState();
+  }
+  
+  // TODO - switch to MotorControl interface instead of AbstractMotor
+  public double calcOutput(AbstractMotor mc) {
+    double in = mc.getPowerLevel();
+    
+    // FIXME - an optimization would be to "set" the values of the MotorControl
+    // if they are null "once" in the attach - so this doesn't have to be done each 
+    // calc
+    double minX = (mc.getMinX() == null)? this.minX:mc.getMinX();
+    double maxX = (mc.getMaxX() == null)? this.maxX:mc.getMaxX();
+    double minY = (mc.getMinY() == null)? this.minY:mc.getMinY();
+    double maxY = (mc.getMaxY() == null)? this.maxY:mc.getMaxY();
+    
+    // motor control has highest precedence
+    Double minInput = mc.getMinInput();
+    Double maxInput = mc.getMaxInput();
+    // motor controller has next order of precedence
+    if (minInput == null) {
+      minInput = this.minInput;
+    }
+    if (maxInput == null) {
+      maxInput = this.maxInput;
+    }
+
+    // motor control has highest precedence
+    Double minOutput = mc.getMinOutput();
+    Double maxOutput = mc.getMaxOutput();
+    // motor controller has next order of precedence
+    if (minOutput == null) {
+      minOutput = this.minOutput;
+    }
+    if (maxOutput == null) {
+      maxOutput = this.maxOutput;
+    }
+    
+    if (minInput != null && in < minInput) {
+      log.warn("clipping input {} to {}", in, minInput);
+      in = minInput;
+    }
+    if (maxInput != null && in > maxInput) {
+      log.warn("clipping input {} to {}", in, maxInput);
+      in = maxInput;
+    }
+    
+    double c = minY + ((in - minX) * (maxY - minY)) / (maxX - minX);
+
+    if (minOutput != null && c < minOutput) {
+      log.warn("clipping output {} to {}", c, minOutput);
+      return minOutput;
+    }
+    if (maxOutput != null && c > maxOutput) {
+      log.warn("clipping output {} to {}", c, maxOutput);
+      return maxOutput;
+    }
+    return c;
+  }
+  
+  // FIXME - if kept they should be put in interface
+  public boolean motorDetach(String name) {
+    if (motors.containsKey(name)) {
+      motors.remove(name);
+      return true;
+    }
+    return false;
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMove(String name) {
+    motorMove((MotorControl) Runtime.getService(name));
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMoveTo(String name, Integer position) {
+    error("not implemented");
+  }
+  
+  @Override
```

Comment:
should throw an unimplemented exception or they should have an implementation

Reply:
maybe ... this shows a nice red glowing error bar in the gui, which might be nicer to teaching a new user that continuous dc motors without feedback cannot be movedTo